File: Lua/systemList.lua
require("tableExtensions")

local Systems = {}
local SystemMT = {}
local EventQueue = table.newqueue()

function SystemMT.addSystem(sys)
  table.insert(Systems, sys)
  return sys
end

local function insertSystem(t, i)
  i = i or 1
  if t[i] ~= nil then
    return SystemMT.addSystem(t[i]), insertSystem(t, i + 1)
  end
end

function SystemMT.addSystems(...)
  return insertSystem({...})
end

function SystemMT.removeAllSystems()
  Systems = {}
end

function SystemMT.removeSystem(system)
  assert(is_a(system, require("system")) or type(system) == "number", "Must enter a system or an id number to 'removeSystem.'")
  local newSystems = {}
  local removed = false
  for _, curSystem in pairs(Systems) do
    if curSystem == system or curSystem:getID() == system then
      removed = true
    else
      table.insert(newSystems, curSystem)
    end
  end
  if removed then
    Systems = newSystems
  end
  return removed
end

function SystemMT.hasEvent(eventName, args)
  for ev in table.queuepairs(EventQueue) do
    if ev[1] == eventName and ev[2] == args then
      return true
    end
  end
  return false
end

function SystemMT.pushEvent(eventName, args)
  SystemMT.pushEventRight(eventName, args)
end

function SystemMT.pushEventRight(eventName, args)
  table.pushright(EventQueue, {eventName, args})
end

function SystemMT.pushEventLeft(eventName, args)
  table.pushleft(EventQueue, {eventName, args})
end

function SystemMT.flushEvents()
  if #Systems == 0 then return end

  local size = table.queuesize(EventQueue)
  if size == 0 then return end
  local handled = 0

  local event, name, args
  while handled < size and table.queuesize(EventQueue) > 0 do
    event = table.popleft(EventQueue)
    name, args = event[1], event[2]
    for _, system in ipairs(Systems) do
      if system:isEnabled() or (args and args.ignoreEnabled) then
        system:dispatchEvent(name, args)
      end
    end
    handled = handled + 1
  end
end

function SystemMT.getSystem(value)
  for _, system in ipairs(Systems) do
    if system:getID() == tonumber(value) or system:getName() == value then
      return system
    end
  end
end

local function runOnSystem(t, i, f, ...)
  i = i or 1
  if t[i] ~= nil then
    return f(t[i], ...), runOnSystem(t, i + 1, f, ...)
  end
end

function SystemMT.forEachSystem(func, ...)
  return runOnSystem(Systems, 1, func, ...)
end

function SystemMT.getCount()
  return #Systems
end

SystemMT.__index = SystemMT
return setmetatable({}, SystemMT)

File: Javascript/systemList.js
define(['ecsevent', 'system'], function(ECSEvent, System)
{
  class SystemList
  {
    constructor()
    {
      this.systems = [];
    }

    addSystem(system, i)
    {
      if(typeof system == "string")
      {
        system = new System(system);
      }

      i = i || 0;
      this.systems.splice(i, 0, system);
      return system;
    }

    hasSystem(system)
    {
      return this.systems.includes(system);
    }

    removeSystem(system)
    {
      for(var i = 0; i < this.systems.length; ++i)
      {
        if(this.systems[i].id == system.id)
        {
          this.systems.splice(i, 1);
          return true;
        }
      }
      return false;
    }

    removeAllSystems()
    {
      this.systems = [];
    }

    pushEvent(eventName, args)
    {
      var ev;
      if(typeof eventName === "string")
      {
        ev = new ECSEvent(eventName, args);
      }
      else if(eventName instanceof ECSEvent)
      {
        ev = eventName;
      }
      else
      {
        throw "Bad event: " + eventName;
      }

      for(var system of this.systems)
      {
        system.pushEvent(ev);
      }
    }

    flushEvents()
    {
      var count = 0;
      for(var system of this.systems)
      {
        count += system.flushEvents();
      }
      return count;
    }

    getSystem(value)
    {
      for(var i = 0; i < this.systems.length; ++i)
      {
        var sys = this.systems[i];
        if(sys.id == value || sys.name == value)
        {
          return sys;
        }
      }
    }

    find(id)
    {
      for(var system of this.systems)
      {
        if(system.id == id)
        {
          return system;
        }

        for(var entity of system)
        {
          if(entity.id == id)
          {
            return entity;
          }
          for(var component of entity)
          {
            if(component.id == id)
            {
              return component;
            }
          }
        }
      }
    }

    forEachSystem(func, args)
    {
      var results = [];
      for(var i = 0; i < this.systems.length; ++i)
      {
        var sys = this.systems[i];
        results.push(func(sys, args));
      }
      return results;
    }

    get systemCount()
    {
      return this.systems.length;
    }

    toSimpleObject()
    {
      var data = [];

      for(var i = 0; i < this.systems.length; ++i)
      {
        var system = this.systems[i];
        data.push(system.toSimpleObject());
      }

      return data;
    }
  }

  return new SystemList();
});
